doc/architecture.md

```diff
@@ -0,0 +1,967 @@
+has serious drawbacks:
+
+- Users may reasonably desire to manipulate remotes using straight Git,
+  bypassing our tools, invalidating our invariants, and creating difficult to
+  diagnose and repair situations.
+- Developers will naturally add remotes for the forks of other developers that
+  they collaborate with.  The requirement to fetch every remote in every
+  sub-repo (even if done in parallel) could cause performance problems.
+- Even using our tools as designed, developers may easily create invalid,
+  difficult-to-recover-from situations.  For example, if a developer makes a
+  local branch from a remote branch, then removes the remote from which that
+  branch came, they may not be able to find the needed commits when opening
+  sub-repos:
+
+```
+'-------------`
+| meta-repo   |
+| - - - - - - |
+| origin      |
+|   master    |
+|    [m1]     |
+| jill        |
+|   master    |
+|    [j2->m1] |
+| - - - - - - |
+| *master     |
+|   [j2]      |
+`-------------,
+```
+
+```bash
+$ git meta remote rm jill
+```
+
+```
+'-------------`
+| meta-repo   |
+| - - - - - - |
+| origin      |
+|   master    |
+|    [m1]     |
+| - - - - - - |
+| *master     |
+|   [j2]      |
+`-------------,
+```
+
+Now even `git meta open` will be unable to initialize the submodule `a` because
+Bob's `master` branch references a commit in it that cannot be found; we no
+longer have any knowledge that Jill's fork exists.
+
+## Conclusions
+
+1. It is not generally possible to synchronize or validate updates to a ref
+   with the same name across many repositories.  Therefore, we treat ref names
+   in only meta-repos as significant; git-meta does not push branches or tags
+   in sub-repos.
+1. We use symbolic-meta-refs as push-targets in sub-repos; as the contents of a
+   symbolic-meta-ref are immutable (a given symbolic-meta-ref can point to only
+   one commit), we are always guaranteed to be able to update them when needed.
+1. Forking sub-repos is potentially expensive, an probably impractical on the
+   server-side, and creates many complications on the client-side.  Therefore,
+   we allow forking of only meta-repos; each sub-repo has a single namespace
+   for the entire mono-repo.
+1. However, since git-meta does not push branch or tag names to sub-repos, and
+   synthetic-meta-branches are plain refs that must be explicitly fetched,
+   there will not be a problem with name-explosion in sub-repos.
+
+# Performance
+
+At a minimum, users working in a mono-repo must download the meta-repo and all
+sub-repos containing code that they require to work.
+
+There is a commit in the meta-repo for every change made in the organization,
+so the number of commits in the history of the meta-repo may be very large.
+However, the information contained in each commit is relatively small,
+generally indicating only changes to submodule pointers.  Furthermore, the
+on-disk (checked out) rendering of the meta-repo is also small, being only a
```

Comment:
Perhaps mention shallow clones - this significantly speeds up our cloning since our cost is the history, not the working directory.